playbooks/files/backup/backup.sh

```diff
@@ -0,0 +1,150 @@
+#   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+
+BACKUP_FROM_DIR="${BACKUP_FROM_DIR:-/tmp/bk-snap}"
+
+get_backup_location(){
+  # Get backup location for loop device name
+  local loop_device_name=$1
+  echo $BACKUP_FROM_DIR/$(echo $loop_device_name | sed 's/--snap//')
+}
+
+get_volumes(){
+  # Get LVM volumes that are not snapshots
+  lvs --noheadings -o vg_name,lv_name,lv_attr | awk '$3 ~ /^[-oV]/ {print $1"/"$2}'
+}
+
+is_partition_table(){
+  # Exit with 0 if the parameter is a device with a partition table, 1 otherwise.
+  blkid $1 | grep -q " PTTYPE="
+}
+
+is_filesystem(){
+  # Exit with 0 if the parameter is a device with a filesystem, 1 otherwise.
+  blkid $1 | grep -q " TYPE="
+}
+
+teardown(){
+  # Remove MBR and partition tables backup
+  lsblk -l -o TYPE,NAME | grep "^disk " | while read _ harddrive
+  do
+    rm -f $BACKUP_FROM_DIR/$harddrive.mbr
+    rm -f $BACKUP_FROM_DIR/$harddrive.sfdisk
+  done
+  # Umount boot and EFI
+  for mount_point in boot boot_efi
+  do
+    mountpoint -q $BACKUP_FROM_DIR/$mount_point &&
+      umount $BACKUP_FROM_DIR/$mount_point
+  done
+  # Umount filesystems and remove snapshots
+  get_volumes | while read volume
+  do
+    test ! -b /dev/$volume-snap &&
+      continue
+    # The created snapshot can contain a disk with partitions that are mounted or
+    # dumped. These partitions will be unmounted or dumps will be deleted.
+    # Otherwise, if the created snapshot does not contain a disk with partitions, the
+    # snapshot will be unmounted or dump will be deleted.
+    if is_partition_table /dev/$volume-snap
+    then
+      kpartx -l /dev/$volume-snap | while read loop_device _
+      do
+        findmnt /dev/mapper/$loop_device -o TARGET | while read mount_point
+        do
+	  umount /dev/mapper/$loop_device
+	  rmdir $mount_point
+	done
+	test -f $(get_backup_location $loop_device) &&
+          rm -f $(get_backup_location $loop_device)
+      done
+      kpartx -d /dev/$volume-snap
+      rm -f $BACKUP_FROM_DIR/$volume.mbr
+      rm -f $BACKUP_FROM_DIR/$volume.sfdisk
+    else
+      mountpoint -q $BACKUP_FROM_DIR/$volume &&
+        umount $BACKUP_FROM_DIR/$volume &&
+	  rmdir $BACKUP_FROM_DIR/$volume
+      test -f $BACKUP_FROM_DIR/$volume &&
+        rm $BACKUP_FROM_DIR/$volume
+    fi
+    /usr/sbin/lvremove --yes $volume-snap
+  done
+}
+
+test ! "${RESTIC_REPOSITORY+x}" -a ! "${RESTIC_REPOSITORY_FILE+x}" &&
+  echo "ERROR: Environment variable RESTIC_REPOSITORY or RESTIC_REPOSITORY_FILE should be defined" > /dev/stderr &&
+    exit 1
+
+test ! "${RESTIC_PASSWORD+x}" -a ! "${RESTIC_PASSWORD_COMMAND+x}" -a ! "${RESTIC_PASSWORD_FILE+x}" &&
+  echo "ERROR: Environment variable RESTIC_PASSWORD, RESTIC_PASSWORD_COMMAND or RESIC_PASSWORD_FILE should be defined" > /dev/stderr &&
+    exit 1
+
+trap teardown HUP INT QUIT ABRT ALRM TERM KILL
```

Comment:
you can just trap "EXIT" instead so the function is called regardless of how the script ended (plus, you cannot trap SIGKILL, as the kernel kills the process immediately)
```suggestion
trap teardown EXIT
```

Reply:
Thanks!

Reply:
Should we add INT to catch Control-C when running on interactive console?